### PR TITLE
add memory cell and rework logic construction

### PR DESCRIPTION
**Pull request by deltanedas**
## About the PR
- made it possible to deconstruct and change types of logic items
  e.g. crowbar a logic gate, add glass and lv to make it an edge detector, crowbar and screw to deconstruct completely
- added memory cell
  when the enabled port is high, it sets its value to the input port's value. it always outputs its current value. when enabled is low input port is ignored.
- all finished logic item sprites have terminals :)
  
you can make memory cell with 3 steel a capacitor and 2 lv cable

## Why / Balance
fuck airlock memory i hate it

this has no network overhead of doors and is computationally cheaper than making the equivalent with a load of gates

things you can do:
- obviously, ram for funny computer
- replace using airlock to store state with a dedicated item
- makes proper cycling airlock controllers much easier because you can store state of which side wants to be open and update it with buttons
- more but i have no imagination

## Technical details
its pretty simple

## Media
in this scenario the switch at the top is the input for both cells
each button pulses the enabled port for its respective cell
cell shows its output with its light bulb

https://github.com/space-wizards/space-station-14/assets/39013340/0f2a1618-285f-45bf-815f-793c23a2b7a0

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
no

**Changelog**
:cl:
- add: Added Memory Cells for storing logic signals persistently.

### CODE REVIEW

**Comment by notquitehadouken** [comment]
the one person who uses logic gates (me) when logic gates get updated:

**Comment by deltanedas** [comment]
i use them too pro :trollface:

**Comment by FungiFellow** [comment]
Nice, my shittles grow ever stronger.

File: Content.Server/DeviceLinking/Systems/MemoryCellSystem.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    {
+        base.Update(deltaTime);
+
+        var query = EntityQueryEnumerator<MemoryCellComponent, DeviceLinkSourceComponent>();
+        while (query.MoveNext(out var uid, out var comp, out var source))
+        {
+            if (comp.InputState == SignalState.Momentary)
+                comp.InputState = SignalState.Low;
+            if (comp.EnableState == SignalState.Momentary)
+                comp.EnableState = SignalState.Low;
```
Comment (Partmedia) [inline Content.Server/DeviceLinking/Systems/MemoryCellSystem.cs:33]: Shouldn't a momentary enable trigger an instantaneous update (capturing input -> output)?

**Review by Partmedia — COMMENTED** [review]
Just to be clear, this "memory cell" is just a "D latch" right? It may be better to name this what this is actually named, to make Googling it easier.

File: Content.Server/DeviceLinking/Systems/MemoryCellSystem.cs (hunk 1 above)
Comment (deltanedas) [reply Content.Server/DeviceLinking/Systems/MemoryCellSystem.cs:33]: probably should but not sure how to handle it in UpdateOutput

**Comment by mirrorcult** [comment]
really feel like with all these logic gate objects + future mechcomp stuff probably getting added we should have like a circuit board item you can just place X amount of these on and connect them freely to save space + have presets for mappers

**Comment by deltanedas** [comment]
fish game circuit box

would be worth having an abstract circuit thing that can be processed outside of items and not using device net

then the items are an interface for them and device network

1 entity with 1 (+engine) component inside circuit box container

**Comment by Emisse** [comment]
Closed due to feature freeze May 10th-June 14th. Comment to have it reopen after this.

**Comment by deltanedas** [comment]
open

**Comment by Partmedia** [comment]
Could you please expand your PR description with some examples of things that people would want to make with this?

**Comment by deltanedas** [comment]
done i guess

**Comment by Partmedia** [comment]
Based on the description you added, this doesn't seem like a viable option for improving airlock controllers. Are you expecting mappers to make a copy of this thing you made for every airlock?

It would be better to improve the airlock signal ports and/or have a dedicated airlock controller. Or, just improve the airlocks themselves.

**Comment by deltanedas** [comment]
> Based on the description you added, this doesn't seem like a viable option for improving airlock controllers. Are you expecting mappers to make a copy of this thing you made for every airlock?

for use ingame when making a perfect shittle, mappers can use tiny fans or since air isnt limited theres no reason to not just autobolt and call it a day.

if it was actually implemented a fish game circuit box would be needed to avoid spamming the same few items over and over, then just have an airlock controlled circuit box and link inputs and outputs easy peasy. but thats in the future.

**Comment by Partmedia** [comment]
Based on the feedback above, I don't think this is sufficiently useful to add without implementing the "fish game circuit box".

If you can think of other good reasonable uses for this, that may change things.

**Comment by deltanedas** [comment]
xor and nor nand xnor are all useless for most people, why have them then

**Comment by juliangiebel** [comment]
I'd rather have this in than see people use airlocks for circuits. Even if it's not done a lot.
In the future logic ICs should be moved to being placed inside a single entity like it was suggested already but who knows when that'll be a thing so I think the memory IC is a decent stopgap.

**Comment by lzk228** [comment]
we should have our own mumbo jumbo

**Comment by deltanedas** [comment]
newgen sethbling here

**Pull request merged**